File: python/basics/linkedlist/21_merge_sorted_list.py
'''
https://leetcode.com/problems/merge-two-sorted-lists/

our typical merge loop..for both valid lists

#meta
'''
from linkedlist import ListNode

def merge_sorted_list(list1:ListNode, list2:ListNode)->ListNode:
    node1=list1
    node2=list2
    result = ListNode(0)
    resultNode = result
    while node1 and node2:
        val = -1
        if node1.val < node2.val:
            val = node1.val
            node1=node1.next
        else:
            val = node2.val
            node2=node2.next

        resultNode.next = ListNode(val)
        resultNode = resultNode.next
    node = node1
    if not node:
        node = node2
    resultNode.next = node
    return result.next



A = ListNode(1)
B = ListNode(2)
C = ListNode(4)
A.next = B
B.next = C
print( "traverse: " + A.traverse())

A1 = ListNode(1)
B1 = ListNode(3)
C1 = ListNode(4)
A1.next = B1
B1.next = C1
print( "traverse: " + A1.traverse())


node = merge_sorted_list(A, A1)
if node:
    print( "traverse node: " + node.traverse())
#####################
A = None
#print( "traverse: " + A.traverse())

A1 = None
#print( "traverse: " + A1.traverse())


node = merge_sorted_list(A, A1)
if node:
    print( "traverse node: " + node.traverse())

#######################################
A = None
#print( "traverse: " + A.traverse())

A1 = ListNode(0)
print( "merged: " + A1.traverse())

node = merge_sorted_list(A, A1)
if node:
    print( "merged: " + node.traverse())
#########################

File: problem_solving/linkedlist/cpp/21_merge_sorted_list.cpp
/*
'''
https://leetcode.com/problems/merge-two-sorted-lists/

our typical merge loop..for both valid lists


#meta
'''
*/

#include<iostream>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
using namespace std;


ListNode* merge_ll(ListNode* l1, ListNode* l2) {
    ListNode* result = new ListNode(0);
    auto node1 = l1;
    auto node2 = l2;
    auto node_result = result;
    while(node1 && node2) {
        if(node1->val <= node2->val) {
            node_result->next = node1;
            node1 = node1->next;
        } else {
            node_result->next = node2;
            node2 = node2->next;
        }
        node_result = node_result->next;
    }
    if(node1) {
        node_result->next = node1;
    } else {
        node_result->next = node2;
    }
    return result->next;
}


int main() {
    ListNode* a = new ListNode(1);
    a->next = new ListNode(2);
    a->next->next = new ListNode(4);
    print_linked_list(a);

    ListNode* a1 = new ListNode(1);
    a1->next = new ListNode(3);
    a1->next->next = new ListNode(4);
    print_linked_list(a1);
    auto result = merge_ll(a, a1);
    print_linked_list(result);
    return 0;
}


